### PR TITLE
Add ecpg tests to the meson

### PR DESCRIPTION
**Pull request by nbyavuz**
This pr adds ecpg tests.

Steps are:

1. Create `*.c and executables files` from `*.pgc` files (using `ecpg`), done in `test/{test_directory}/meson.build` files.
2. Copy `expected` files from source directory to the build directory.
3. Edit `pg_regress_ecpg.c` so it will try to read `expected files` from `inputdir` if files don't exist in `outputdir`.
4. Create `pg_regress_ecpg` executable in `ecpg/test/meson.build` file.
5. Run tests with `pg_regress_ecpg`.

### CODE REVIEW

File: src/interfaces/ecpg/test/pgtypeslib/meson.build [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+]
+
+foreach pgc_file: pgc_files
+  exe_input = custom_target(pgc_file,
+    input: '@0@.pgc'.format(pgc_file),
+    output: '@0@.c'.format(pgc_file),
+    command: [
+      ecpg_exe, 
+      '--regression',
+      '-I' + meson.current_source_dir(),
```
Comment (anarazel) [inline src/interfaces/ecpg/test/pgtypeslib/meson.build:16]: The source-dir reference should probably replaced by `@CURRENT_SOURCE_DIR@` and `@SOURCE_ROOT@`.

The output file `.format()` could be replaced with `@PLAINNAME@`.

With that the entire command shouldn't differ between the directories. That should allow nearly all the arguments to be specified as a single kwargs: ecpg_preproc_test_kwargs?

File: src/interfaces/ecpg/test/pgtypeslib/meson.build [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      '-o', '@OUTPUT@',
+      '@INPUT@'
+    ],
+  )
+
+  executable(pgc_file,
+    exe_input,
+    dependencies: [frontend_code, libpq],
+    include_directories: [ecpg_inc],
+    link_with: [ecpglib, ecpg_compat, ecpg_pgtypes],
```
Comment (anarazel) [inline src/interfaces/ecpg/test/pgtypeslib/meson.build:27]: It'd be good to mark this as "build_by_default: false". It might again be a good argument to replace most of the arguments with a kwargs: ecpg_test_exec_kw to avoid repeating all of this / needing to adjust multiple places when changing?

File: src/interfaces/ecpg/test/preproc/meson.build [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  input: 'strings.pgc',
+  output: 'strings.c',
+  command: [
+    ecpg_exe, 
+    '--regression',
+    '-I' + meson.current_source_dir(),
+    '-I' + meson.current_source_dir() + '/../../include/',
+    '-i', '-o', '@OUTPUT@', '@INPUT@'
+  ],
+)
```
Comment (anarazel) [inline src/interfaces/ecpg/test/preproc/meson.build:75]: Why are these not using the loop approach above?

File: src/interfaces/ecpg/test/meson.build [hunk 4]
```diff
@@ -0,0 +1,54 @@
+# TODO: ECPG is not enabled for windows yet
+if host_machine.system() == 'windows'
+  subdir_done()
+endif
```
Comment (anarazel) [inline src/interfaces/ecpg/test/meson.build:4]: Why is this needed - shouldn't the equivalent bit in src/interfaces/ecpg/meson.build already prevent getting here?

File: src/interfaces/ecpg/test/expected/meson.build [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  'thread-thread_implicit_2.stdout',
+]
+
+foreach copy_file: copy_files_list
+  configure_file(
+    output: '@0@'.format(copy_file),
+    input: '@0@'.format(copy_file),
+    copy : true
+  )
+endforeach
```
Comment (anarazel) [inline src/interfaces/ecpg/test/expected/meson.build:207]: I think it'd be strongly preferrable to avoid the need for this copying by adjusting the test infrastructure to find the file in their original place. As far as I can see ecpg's pg_regress already has --inputdir,  right?

Comment (eli-schwartz) [reply src/interfaces/ecpg/test/expected/meson.build:207]: +1

Depending on the project I've sometimes also been able to get it to find files by setting the `workdir` kwarg to run in the source directory instead of the build root

File: src/interfaces/ecpg/test/pgtypeslib/meson.build (hunk 2 above)
Comment (nbyavuz) [reply src/interfaces/ecpg/test/pgtypeslib/meson.build:27]: These executables are needed for running ecpg tests. If I mark them as "build_by_default: false", tests fail.

File: src/interfaces/ecpg/test/pgtypeslib/meson.build (hunk 1 above)
Comment (nbyavuz) [reply src/interfaces/ecpg/test/pgtypeslib/meson.build:16]: These custom targets have different flags in commands like `'-C', 'INFORMIX'`, `'-C', 'ORACLE'`, `'-r', 'prepare'`, `'-i'`, instead I can create variable(ecpg_preproc_test_command) for same parts of the commands.

File: src/interfaces/ecpg/test/preproc/meson.build (hunk 3 above)
Comment (nbyavuz) [reply src/interfaces/ecpg/test/preproc/meson.build:75]: That has extra `-i` flag.

File: src/interfaces/ecpg/test/expected/meson.build (hunk 5 above)
Comment (nbyavuz) [reply src/interfaces/ecpg/test/expected/meson.build:207]: The problem is, when we set --inputdir; pg_regress looks for expected files and .c, exe files in the --inputdir directory. Expected files are in the source directory but .c and exe files are in the build directory(since these are generated by using executable and custom_target). 

Maybe we can add something like --expecteddir to the pg_regress?

File: src/interfaces/ecpg/test/pgtypeslib/meson.build (hunk 2 above)
Comment (eli-schwartz) [reply src/interfaces/ecpg/test/pgtypeslib/meson.build]: If they are needed by the tests, then they should be added as `depends: ...` to the `test()` definition.

Comment (nbyavuz) [reply src/interfaces/ecpg/test/pgtypeslib/meson.build]: @eli-schwartz, we are generating .c files from .pgc files, then generating executables from these .c files. Executables and .c files are both needed by the tests. So, should `custom targets(for generating .c files from .pgc files)` be added as `depends` to the `executables()` definition?

Comment (eli-schwartz) [reply src/interfaces/ecpg/test/pgtypeslib/meson.build]: Adding the executable as a dependency of the test ensures that it is built before tuning the test (even if `build_by_default: false`).

The C files are already dependencies of the executable because they are in the sources, so they don't need to be added to the test also.

Comment (nbyavuz) [reply src/interfaces/ecpg/test/pgtypeslib/meson.build]: Thanks! I have one more question, as far as I understand these executables will be created if they added as `depends:` to the `test()` right(No need to run tests)? And we are adding these executables as `depends:` to the `test()` as long as possible(We are trying to define all possible tests). So, for it to be useful should we have option like `-Dbuild_tests=true`?

Comment (eli-schwartz) [reply src/interfaces/ecpg/test/pgtypeslib/meson.build]: I'm not sure what you mean by this?

You're adding it to the depends to ensure that correct dependencies are set up, which is important for two reasons:
- you can `git bisect` and run a single test via `meson test test-name` which will not rebuild the entire project, only the dependencies for that specific test
- if you set `build_by_default: false` you need to add it as a test dep to make sure it is built before running the testsuite (let's ignore, for now, the fact that meson has a long-standing bug that test dependencies are re-added to the default target; this will eventually be fixed to only build it when running the test target, at which point an existing, well-defined project will immediately benefit)

Comment (nbyavuz) [reply src/interfaces/ecpg/test/pgtypeslib/meson.build]: Thanks! Your answer has been explanatory.

File: src/interfaces/ecpg/test/expected/meson.build (hunk 5 above)
Comment (anarazel) [reply src/interfaces/ecpg/test/expected/meson.build:207]: > The problem is, when we set --inputdir; pg_regress looks for expected files and .c, exe files in the --inputdir directory. Expected files are in the source directory but .c and exe files are in the build directory(since these are generated by using executable and custom_target).

Ah, I see.

> Maybe we can add something like --expecteddir to the pg_regress?

That sounds like a good plan.

File: src/interfaces/ecpg/test/preproc/meson.build (hunk 3 above)
Comment (anarazel) [reply src/interfaces/ecpg/test/preproc/meson.build:59]: You can solve that by doing something like
```
pgc_files = [
 ...
  'array_of_struct' : {'extra_flags': ['-c']},
  'pointer_to_struct' : {'extra_flags': ['-c']},
  'autoprep': {'extra_flags': ['-r', 'prepare']},
  'strings': {'extra_flags: ['-i']},
]
```

Comment (nbyavuz) [reply src/interfaces/ecpg/test/preproc/meson.build]: Updated.

File: src/interfaces/ecpg/test/expected/meson.build (hunk 5 above)
Comment (nbyavuz) [reply src/interfaces/ecpg/test/expected/meson.build]: Updated.

**Review by anarazel — COMMENTED** [review]
Could you split this into one commit for the changes to pg_regress_ecpg.c and the rest?

File: src/interfaces/ecpg/test/meson.build [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  'dependencies': [frontend_code, libpq],
+  'include_directories': [ecpg_inc],
+  'link_with': [ecpglib, ecpg_compat, ecpg_pgtypes],
+  'build_by_default': false,
+}
+
+ecpg_preproc_test_command_start = [
+  ecpg_exe, 
+  '--regression',
+  '-I' + '@CURRENT_SOURCE_DIR@',
```
[the remark this answers is not in the document]
Comment (nbyavuz) [reply src/interfaces/ecpg/test/meson.build]: Done.

File: src/interfaces/ecpg/test/preproc/meson.build [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+pgc_extra_flags = {
+  'array_of_struct': ['-c'],
+  'pointer_to_struct' : ['-c'],
+  'autoprep': ['-r', 'prepare'],
+  'strings': ['-i'],
+}
+
+foreach pgc_file: pgc_files
+  exe_input = custom_target(pgc_file,
+    input: '@0@.pgc'.format(pgc_file),
```
Comment (anarazel) [inline src/interfaces/ecpg/test/preproc/meson.build:27]: Seems like we could just use have pgc_files be the actual filenames, instead of the .format() thing?

File: src/interfaces/ecpg/test/meson.build [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+ecpg_regress_args = [
+  '--dbname=ecpg1_regression,ecpg2_regression',
+  '--create-role=regress_ecpg_user1,regress_ecpg_user2',
+  '--encoding=SQL_ASCII',
+]
+
+tests += {
+  'name': 'ecpg',
+  'sd': meson.current_build_dir(),
```
Comment (anarazel) [inline src/interfaces/ecpg/test/meson.build:60]: This should be the source dir, I think?

File: src/interfaces/ecpg/test/meson.build [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  'bd': meson.current_build_dir(),
+  'regress': {
+    'test_runner': pg_regress_ecpg,
+    'expecteddir': meson.current_source_dir(),
```
Comment (anarazel) [inline src/interfaces/ecpg/test/meson.build:64]: Maybe I'm just slow right now. But how does this differ from --inputdir? That's also set to the source directory? Oh, I see - you intentionally set 'sd' to something other than the source directory? I don't think that's good...

Comment (nbyavuz) [reply src/interfaces/ecpg/test/meson.build:64]: Yes, it was intentional. But since generated .c files and executables(in that case, these are input files) are in the build directory, I set sd as a build directory. Other way could be generating .c files and executables in the source directory but that also seems wrong to me. Do you have any other idea?

Current logic is:
.pgc files are in the source directory (we are getting these as an input at custom_target())
expected files are in the build directory (expecteddir is added)
.c and executables in the build directory (I set sd as a build directory)

Comment (nbyavuz) [reply src/interfaces/ecpg/test/meson.build:64]: There can be small confusion, source files are not generated by setting the source directory to the build directory. Source files of ecpg_test are taken from build directory.

There are two steps:
1. Generating .c and executables, which is done by setting custom_target and executables. We are getting these .pgc files from source directory; then create .c and executables in the build directory.
2. These executables and .c files are used in ecpg regression test. ecpg_test needs to read these files and since these .c and executables are in the build directory, I set ecpg_test's sd as a build directory.

Comment (anarazel) [reply src/interfaces/ecpg/test/meson.build:64]: > There can be small confusion, source files are not generated by setting the source directory to the build directory. Source files of ecpg_test are taken from build directory.

My point is that you're setting `'sd'` to `meson.current_build_dir()`, but `'sd'` stands for source directory. If you need to change where the test gets the input from or such, you should override that parameter, instead of "lying" about 'sd'.

File: src/interfaces/ecpg/test/preproc/meson.build (hunk 7 above)
Comment (nbyavuz) [reply src/interfaces/ecpg/test/preproc/meson.build:27]: If I apply that change,
```
executable(pgc_file,
...
)
```

 will be a problem.

File: src/interfaces/ecpg/test/meson.build (hunk 9 above)
Comment (nbyavuz) [reply src/interfaces/ecpg/test/meson.build:64]: Updated.

**Comment by anarazel** [comment]
Yay, thanks!
